dbms/src/Interpreters/Expand2.cpp

```diff
@@ -0,0 +1,107 @@
+{
+    /// step1: clone a new block
+    ColumnsWithTypeAndName cloned_columns_and_type;
+    auto num_columns = block_cache.columns();
+    MutableColumns res(num_columns);
+    for (size_t i = 0; i < num_columns; ++i)
+    {
+        res[i] = block_cache.getColumns()[i]->cloneResized(block_cache.rows());
+    }
+    auto origin_column_and_types = block_cache.getColumnsWithTypeAndName();
+    for (size_t i = 0; i < num_columns; ++i)
+    {
+        cloned_columns_and_type.emplace_back(std::move(res[i]), origin_column_and_types[i].type, origin_column_and_types[i].name, origin_column_and_types[i].column_id, origin_column_and_types[i].default_value);
+    }
+    Block cloned_block(cloned_columns_and_type);
+
+    /// step2: execute a new block.
+    auto ith_projection = leveled_projections_actions[i_th_project];
+    ith_projection->execute(cloned_block);
+
+    /// step3: organize a new block with unified alias name.
+    Block new_block;
+    auto names_with_alias = leveled_alias_projections[i_th_project];
+    for (auto & one_alias : names_with_alias)
+    {
+        const std::string & name = one_alias.first;
+        const std::string & alias = one_alias.second;
+        ColumnWithTypeAndName column = cloned_block.getByName(name);
+        if (!alias.empty())
+            column.name = alias;
+        new_block.insert(std::move(column));
+    }
+
+    /// step4: unfold the constant column which is not meaningful in global scope.
+    for (size_t i = 0; i < new_block.getColumnsWithTypeAndName().size(); i++)
+    {
+        auto col = new_block.getColumnsWithTypeAndName()[i];
+        if (col.column->isColumnConst())
+        {
+            // if it's a new literal constant column, unfold it, and if it's not an origin constant column, unfold it.
+            // eg: grouping id projection(1) and grouping set column projection(null)
+            if (!block_cache.getByName(col.name).column->isColumnConst())
+                new_block.safeGetByPosition(i).column = col.column->convertToFullColumnIfConst();
+        }
+    }
+    return new_block;
+}
+
+String Expand2::getLevelProjectionDes() const
+{
+    FmtBuffer buffer;
+    buffer.append("[");
+    buffer.joinStr(
+        leveled_projections_actions.begin(),
+        leveled_projections_actions.end(),
+        [](const auto & item, FmtBuffer & buf) {
+            // for every level-projection, make it as string too.
+            buf.append("[");
+            buf.joinStr(
+                item->getActions().begin(),
+                item->getActions().end(),
+                [](const auto & item, FmtBuffer & buff) { buff.append(item.toString()); },
+                ",");
+            buf.append("]");
+        },
+        ";\n");
+    buffer.append("]");
+    return buffer.toString();
+}
+
+size_t Expand2::getLevelProjectionNum() const
+{
+    return leveled_projections_actions.size();
+}
+
+ExpressionActionsPtr & Expand2::getBeforeExpandActions()
```

Comment:
We should avoid to return reference as possible as we can and the type is already a smart pointer, so I think return `ExpressionActionsPtr` is enough.

Reply:
![image](https://github.com/pingcap/tiflash/assets/15825830/379b4cba-0a19-4826-8dbb-ffa11647bdfc)
actually, we wanna pass it to this function which receive a & ref of expressionActions, a ExpressionActionsPtr & will be more clear. Otherwise, we need a make a var to get the temporary value of what getBeforeExpandActions returned.